src/EGraphs/egraph.jl

```diff
@@ -223,38 +223,20 @@ Returns the canonical e-class id for a given e-class.
 
 @inline Base.getindex(g::EGraph, i::Id) = g.classes[IdKey(find(g, i))]
 
-# function canonicalize(g::EGraph, n::VecExpr)::VecExpr
-#   if !v_isexpr(n)
-#     v_hash!(n)
-#     return n
-#   end
-#   l = v_arity(n)
-#   new_n = v_new(l)
-#   v_set_flag!(new_n, v_flags(n))
-#   v_set_head!(new_n, v_head(n))
-#   for i in v_children_range(n)
-#     @inbounds new_n[i] = find(g, n[i])
-#   end
-#   v_hash!(new_n)
-#   new_n
-# end
-
 function canonicalize!(g::EGraph, n::VecExpr)
-  v_isexpr(n) || @goto ret
-  for i in (VECEXPR_META_LENGTH + 1):length(n)
-    @inbounds n[i] = find(g, n[i])
+  if v_isexpr(n)
+    for i in (VECEXPR_META_LENGTH + 1):length(n)
+      @inbounds n[i] = find(g, n[i])
+    end
   end
-  v_unset_hash!(n)
-  @label ret
-  v_hash!(n)
```

Comment:
Yes, you are right, the main issue is that VecExpr must not be updated after they have been added to memo. Caching of hash values is an independent concern. 
I did a quick analysis, in which I checked how often the cached values are actually used and I saw only a small usage factor. I'll redo the analysis more carefully and post the result here. 

The cached value can make up 15% to 20% of the memory required for VecExpr.

Reply:
Using the annotations in https://github.com/gkronber/Metatheory.jl/tree/count_vecexpr_hash_calls
```julia
julia> using Metatheory
Precompiling Metatheory
  1 dependency successfully precompiled in 2 seconds. 6 already precompiled.

julia> include("benchmark/benchmarks.jl")
Benchmark(evals=1, seconds=5.0, samples=10000)

julia> run(SUITE)
[...]

julia> Metatheory.VecExprModule.vexpr_created
12355173
julia> Metatheory.VecExprModule.v_copy_calls
5247759
julia> Metatheory.VecExprModule.v_new_calls
3975935
julia> Metatheory.VecExprModule.unset_hash_calls
41556639
julia> Metatheory.VecExprModule.hash_calls
93167599
julia> Metatheory.VecExprModule.cached_hash_computation
40712170
julia> Metatheory.VecExprModule.cached_hash_access
1205819
julia> Metatheory.EGraphs.memo_lookups
31998547
julia> Metatheory.EGraphs.memo_add
15708295
```

In this run of the benchmarks:
- 12mio ```VecExpr``` objects are constructed (some via v_new, some via copy, most via direct constructor calls)
- 32mio lookups are done in memo (note that most of them call the hash function twice via ```haskey(n) && g.memo[n]```). 
- 15mio times ```g.memo[n]``` is set (memo_add)
- memo lookups and adding to memo cause 93mio calls to hash(n::VecExpr)
- ```v_hash!``` is called 42mio times (40.7mio times the hash value is calculated, 1.2mio times the cached value is returned)
- -> only a bit more than half of the hash calls can use the cached value
- if instead of ```haskey(n) && g.memo[n]``` we use ```get``` we should be able to reduce the number of hash calls significantly (exact number to be added in a later edit).

Reply:
> VecExpr must not be updated after they have been added to memo.  

I think this happens as well in `egg` and is part of the algorithm. We should ask the egg community how they are doing it.

Reply:
In egg they are careful to clone nodes before adding them to memo.